File: src/bin/day9.rs
//! Day 9: Disk Fragmenter
//!
//! <https://adventofcode.com/2024/day/9>

use std::fmt::Display;
use aoc_2024::Solution;

pub struct Day9;

impl Solution for Day9 {
    const NAME: &'static str = "Disk Fragmenter";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        let mut disk = Vec::new();
        let mut num = 0;

        for (i, count) in inp
            .to_string()
            .bytes()
            .enumerate()
        {
            let count = count as usize - 48;

            if i % 2 == 0 {
                disk.extend_from_slice(&vec![Some(num); count]);
                num += 1;
            } else {
                disk.extend_from_slice(&vec![None; count]);
            }
        }

        let mut i = 0;

        while i < disk.len() {
            if disk[i].is_none() {
                if let Some(last @ Some(_)) = disk.pop() {
                    disk[i] = last;
                } else {
                    continue;
                }
            }
            i += 1;
        }

        disk.into_iter()
            .enumerate()
            .filter_map(|(i, num)| num.map(|n| n * i))
            .sum()
    }

    fn part_two<T: Display>(&self, inp: T) -> usize {
        let mut disk = Vec::new();
        let mut spaces = Vec::new();
        let mut blocks = Vec::new();

        let mut num = 0usize;

        for (i, count) in inp
            .to_string()
            .bytes()
            .enumerate()
        {
            let count = count as usize - 48;
            let start = disk.len();
            let interval = start..start + count;

            if i % 2 == 0 {
                let block = vec![Some(num); count];

                disk.extend_from_slice(&block);
                blocks.push((block, interval));
                num += 1;
            } else {
                spaces.push(interval);
                disk.extend_from_slice(&vec![None; count]);
            }
        }

        for (block, block_interval) in blocks
            .into_iter()
            .rev()
        {
            let block_size = block_interval.len();

            for space in &mut spaces {
                let space_size = space.len();

                if block_size <= space_size && space.start < block_interval.start {
                    disk[block_interval].fill(None);
                    disk[space.start..space.start + block_size].copy_from_slice(&block);

                    space.start += block_size;
                    break;
                }
            }
        }

        disk.into_iter()
            .enumerate()
            .filter_map(|(i, num)| num.map(|n| n * i))
            .sum()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 6_356_833_654_075);
        assert_eq!(p2, 6_389_911_791_746);
    }
}

fn main() {
    aoc_2024::run_day(9, &Day9);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}

File: aoc-py/solutions/day9.py
"""
Day 9: Disk Fragmenter

https://adventofcode.com/2024/day/9
"""
__all__ = ('Day9',)

from typing import ClassVar

from ..solution import Solution

class Day9(Solution):
    NAME: ClassVar[str] = 'Disk Fragmenter'

    def part_one(self, inp: str) -> int:
        disk = []
        num = 0

        for i, count in enumerate(inp):
            count = int(count)
            if i % 2 == 0:
                disk += [num] * count
                num += 1
            else:
                disk += [None] * count

        i = 0
        while i < len(disk):
            if disk[i] is None:
                if (last := disk.pop()) is None:
                    # simply try again as this continues the loop without increasing `i`
                    continue

                disk[i] = last
            i += 1

        return sum(
            i * num for i, num in enumerate(disk)
        )

    def part_two(self, inp: str) -> int:
        # the parsed disk: '123' -> '0..111'
        disk = []
        # stores a list of the indices interval `[start, stop)` in which each contiguous block of space occupies in the disk
        #
        # i.e. the '..' in '0..111' would be stored as `slice(1, 3)`
        spaces: list[slice] = []
        # stores a list of the literal block and the indices interval `[start, stop)` in which it occupies in the disk
        #
        # i.e. the '111' in '0..111' would be stored as `([1, 1, 1], slice(3, 6))`
        blocks: list[tuple[list[int], slice]] = []

        num = 0

        for i, count in enumerate(inp):
            count = int(count)
            # the start index of the next block will be the length of the existing disk
            start = len(disk)
            interval = slice(start, start + count)

            if i % 2 == 0:
                blocks.append((
                    block := [num] * count,
                    interval,
                ))
                disk += block
                num += 1
            else:
                spaces.append(interval)
                disk += [None] * count

        # try and find a space for all the blocks starting at the end
        for block, block_slice in reversed(blocks):
            block_size = block_slice.stop - block_slice.start

            for i, space in enumerate(spaces):
                space_size = space.stop - space.start

                # found an empty `space` *big enough* and *occurs before* for our `block` to move itself into
                if block_size <= space_size and space.start < block_slice.start:

                    if block_size == space_size:
                        # OPTIMIZATION: remove completely occupied blocks from list
                        del spaces[i]
                    else:
                        # cut off the start of the space that is now occupied by the block
                        spaces[i] = slice(space.start + block_size, space.stop)

                    # replace block's original spot with '...'
                    disk[block_slice] = [None] * block_size
                    # replace the `block_size` sized portion at the start of the `space` with the block itself
                    disk[space.start:space.start + block_size] = block

                    break

        return sum(
            i * num for i, num in enumerate(disk) if num is not None
        )

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 6356833654075
        assert p2 == 6389911791746